lib/rtcpeerconnection/safari.js

```diff
@@ -236,47 +231,58 @@ SafariRTCPeerConnection.prototype.close = function close() {
-  if (this._localStreams.has(stream)) {
-    return;
+// NOTE(mmalavalli): webrtc-adapter has a bug where the "addTrack" shim
+// does not return an RTCRtpSender and returns undefined instead. An issue
+// [https://github.com/webrtc/adapter/issues/714] has been filed. For now,
+// we manually get the RTCRtpSender associated with the added track and
+// return it.
+//
+// UPDATE(mmalavalli): Because we are not delegating to the native
+// RTCPeerConnection#removeTrack(), we have to manually maintain a list of added
+// tracks. So we disable the delegation to the native RTCPeerConnection#addTrack()
+// for now.
+SafariRTCPeerConnection.prototype.addTrack = function addTrack() {
+  var args = [].slice.call(arguments);
+  var track = args[0];
+  var sender = this._senders.get(track.id);
+  if (sender && sender.track === track) {
+    throw new Error('Cannot add MediaStreamTrack [' + track.id + ', '
+      + track.kind + ']: RTCPeerConnection already has it');
   }
-  var tracks = new Set(stream.getTracks());
-  this._peerConnection.getSenders().forEach(function(sender) {
-    if (tracks.has(sender.track)) {
-      tracks.delete(sender.track);
-    }
-  });
-  tracks.forEach(function(track) {
-    try {
-      this.addTrack(track, stream);
-    } catch (error) {
-      // Do nothing.
-    }
-  }, this);
-  this._localStreams.set(stream, new Set(tracks));
-};
+  sender = getActiveSenders(this._peerConnection).get(track.id)
+    || this._peerConnection.addTrack.apply(this._peerConnection, args);
 
-SafariRTCPeerConnection.prototype.removeStream = function removeStream(stream) {
-  // NOTE(mroberts): We can't really remove tracks right now, at least if we
-  // ever want to add them back...
-  //
-  //     https://bugs.webkit.org/show_bug.cgi?id=174327
-  //
-  // var tracks = this._localStreams.get(stream) || new Set();
-  // this.getSenders().forEach(function(sender) {
-  //   if (tracks.has(sender.track)) {
-  //     this.removeTrack(sender);
-  //   }
-  // }, this);
-  this._localStreams.delete(stream);
+  sender = sender || getActiveSenders(this._peerConnection).get(track.id);
+  this._senders.set(track.id, sender);
+  return sender;
 };
 
-SafariRTCPeerConnection.prototype.getLocalStreams = function getLocalStreams() {
-  return Array.from(this._localStreams.keys());
+// NOTE(mroberts): We can't really remove tracks right now, at least if we
+// ever want to add them back...
+//
+//     https://bugs.webkit.org/show_bug.cgi?id=174327
+//
+SafariRTCPeerConnection.prototype.removeTrack = function removeTrack(sender) {
+  var track = sender.track;
+  if (!track) {
+    return;
+  }
+  if (this._isClosed) {
+    throw new Error('Cannot remove MediaStreamTrack [' + track.id +
+      ', ' + track.kind + ']: RTCPeerConnection is closed');
+  }
+  sender = this._senders.get(track.id);
+  if (sender && sender.track === track) {
+    this._senders.set(track.id, new RTCRtpSenderShim(null));
```

Comment:
This seems strange to me. If we are going to support this, I think we should do so without breaking reference equality.

Reply:
@markandrus In case of Firefox and Safari, since we're using the native `RTCRtpSender` objects, we cannot mutate the `.track` member variable. Since we have shimmed `removeTrack`, we mimic its behavior by replacing the `RTCRtpSender` with our shim with a `null` `.track` member.